backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import com.google.gson.Gson;
+import com.google.sps.data.Match;
+import com.google.sps.data.Participant;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/** Class used to find a match in a list of Participants with the most recently added Participant */
+public final class FindMatchQuery {
+
+  private Date date;
+  
+  /** Constructor */
+  public FindMatchQuery() {
+    date = new Date();
+  }
+
+  /** Constructor with manually set date */
+  public FindMatchQuery(Date date) {
+    this.date = date;
+  }
+
+  /** Find match of new participant or add to list of participants, return whether or not match was found 
+    * right after being added
+    */
+  public Match findMatch(List<Participant> participants, Participant newParticipant) {
+    int MAX_DURATION_DIFF = 15; // maximum difference in duration to be compatible
+    int PADDING_TIME = 15; // extra padding time to ensure large enough meeting time block
+    
+    // Compare new participant preferences with others in list to find match
+    for (Participant currParticipant : participants) {
+
+      // Check if participants are looking for similar meeting duration
+      int newDuration = newParticipant.getDuration();
+      int currDuration = currParticipant.getDuration();
+      boolean compatibleDuration = Math.abs(newDuration - currDuration) <= MAX_DURATION_DIFF;
+      int duration = Math.min(newDuration, currDuration);
+      
+      if (!compatibleDuration) {
+        continue;
+      }
+      
+      // Check if participants are both free for that duration + extra
+      long newTimeAvailableUntil = newParticipant.getTimeAvailableUntil();
+      long currTimeAvailableUntil = currParticipant.getTimeAvailableUntil();
+      boolean compatibleTime = date.getTime() + Duration.ofMinutes(duration + PADDING_TIME).toMillis() <=
```

Comment:
boolean compatibleTime = Instant.now().plusMillis(Duration ....).isBefore(......)

Reply:
I don't think I can use Instant.now() here because for testing this class, I have to create a new reference Date. Any suggestions on how to refactor all of the Date/time code to be more consistent and easier to understand/use? All of the date objects I've found are sort of confusing.

Reply:
Using Instant.now() would eliminate the need to use Date in this class and in the test. The test has control over the meeting duration and the `timeAvailableUntil` values which should give enough flexibility to test this statement.

Reply:
I changed FindMeetingQuery to depend on Clock and will change other files later.